src/IO/AzureBlobStorage/copyAzureBlobStorageFile.cpp

```diff
@@ -0,0 +1,324 @@
+
+                {
+                    std::lock_guard lock(bg_tasks_mutex);
+                    task = &bg_tasks.emplace_back();
+                    ++num_added_bg_tasks;
+                }
+
+                /// Notify waiting thread when task finished
+                auto task_finish_notify = [this, task]()
+                {
+                    std::lock_guard lock(bg_tasks_mutex);
+                    task->is_finished = true;
+                    ++num_finished_bg_tasks;
+
+                    /// Notification under mutex is important here.
+                    /// Otherwise, WriteBuffer could be destroyed in between
+                    /// Releasing lock and condvar notification.
+                    bg_tasks_condvar.notify_one();
+                };
+
+                try
+                {
+                    auto read_buffer = std::make_unique<LimitSeekableReadBuffer>(create_read_buffer(), part_offset, part_size);
+                    auto buffer = std::make_unique<StdStreamFromReadBuffer>(std::move(read_buffer), part_size);
+                    task->data = new char[part_size];
+                    task->size = part_size;
+                    buffer->read(task->data,part_size);
+                    task->block_id = getRandomASCIIString(64);
+
+                    schedule([this, task, task_finish_notify]()
+                    {
+                        try
+                        {
+                            processUploadTask(*task);
+                        }
+                        catch (...)
+                        {
+                            task->exception = std::current_exception();
+                        }
+                        task_finish_notify();
+                    }, Priority{});
+                }
+                catch (...)
+                {
+                    task_finish_notify();
+                    throw;
+                }
+            }
+            else
+            {
+                UploadPartTask task;
+                auto read_buffer = std::make_unique<LimitSeekableReadBuffer>(create_read_buffer(), part_offset, part_size);
+                auto buffer = std::make_unique<StdStreamFromReadBuffer>(std::move(read_buffer), part_size);
+                task.data = new char[part_size];
+                buffer->read(task.data,part_size);
+                task.size = part_size;
+                processUploadTask(task);
+                block_ids.emplace_back(task.block_id);
+            }
+        }
+
+        void processUploadTask(UploadPartTask & task)
+        {
+            auto block_id = processUploadPartRequest(task);
+
+            std::lock_guard lock(bg_tasks_mutex); /// Protect bg_tasks from race
+            task.block_id = block_id;
+            LOG_TRACE(log, "Writing part finished. Bucket: {}, Key: {}, block_id: {}, Parts: {}", dest_bucket, dest_key, block_id, bg_tasks.size());
+        }
+
+        String processUploadPartRequest(UploadPartTask & task)
```

Comment:
For S3 `task.block_id` was generated not us but remotely by AWS S3 itself.
For AzureBlobStorage we generate them, so `processUploadPartRequest()` doesn't need to return them.

Reply:
ya I just added that to append to the block_ids list only after that block_id is processed. Any concern if we keep it like this ?